helix-core/src/main/java/org/apache/helix/controller/stages/ResourceComputationStage.java

```diff
@@ -189,56 +189,69 @@ private void processCurrentStates(BaseControllerDataProvider cache,
         String instanceName = instance.getInstanceName();
         String clientSessionId = instance.getEphemeralOwner();
 
-        Map<String, CurrentState> currentStateMap =
-            cache.getCurrentState(instanceName, clientSessionId);
-        if (currentStateMap == null || currentStateMap.size() == 0) {
-          continue;
+        processCurrentStateMap(cache.getCurrentState(instanceName, clientSessionId), resourceMap,
```

Comment:
This is for backward compatibility, right?
Can we create another private method to get the full task current state smartly? So you don't need 2 processCurrentStateMap calls here.

Reply:
We always need to process current states from both paths. In the resource pipeline, we need the regular current states; in the task pipeline, we need task current states + regular for backward compatibility. 
Could you elaborate on the "getting smartly" part? Do you mean: combine the current state maps so there's only one call? That would incur additional memory and computation cost.

Reply:
getCurrentState(boolean isTaskPipeline) {
  currentstates = cache.getCurrentState(...)
  if (isTaskPipeline) {
    filter currentStates for the tasks only
    currentstates.addAll(cache.getTaskCurrentState(...))
  } else {
    filter currentStates for the resources only
  }
  return currentStates
}

Reply:
Right now, it can't be a part of `BaseControllerDataProvider` if that's what you're referring to. `TaskCurrentStateCache` exists in `WorkflowControllerDataProvider` only, and it's for a performance reason since it only needs to be refreshed there, not in the regular pipeline. Without first determining `isTaskPipeline`, we can't parse the base data provider to the workflow data provider.

Of course, we can change how that works. I'll experiment.